lib/Parse/Lexer.cpp

```diff
@@ -1951,36 +1958,35 @@ const char *Lexer::findEndOfCurlyQuoteStringLiteral(const char *Body,
   }
 }
 
-void Lexer::lexRegexLiteral(const char *TokStart) {
+bool Lexer::tryLexRegexLiteral(const char *TokStart) {
   assert(*TokStart == '\'');
 
-  bool HadError = false;
-  while (true) {
-    // Check if we reached the end of the literal without terminating.
-    if (CurPtr >= BufferEnd || *CurPtr == '\n' || *CurPtr == '\r') {
-      diagnose(TokStart, diag::lex_unterminated_regex);
-      return formToken(tok::unknown, TokStart);
-    }
+  // We need to have experimental string processing enabled, and have the
+  // parsing logic for regex literals available.
+  if (!LangOpts.EnableExperimentalStringProcessing || !regexLiteralLexingFn)
+    return false;
 
-    const auto *CharStart = CurPtr;
-    uint32_t CharValue = validateUTF8CharacterAndAdvance(CurPtr, BufferEnd);
-    if (CharValue == ~0U) {
-      diagnose(CharStart, diag::lex_invalid_utf8);
-      HadError = true;
-      continue;
-    }
-    if (CharValue == '\\' && (*CurPtr == '\'' || *CurPtr == '\\')) {
-      // Skip escaped delimiter or \.
-      CurPtr++;
-    } else if (CharValue == '\'') {
-      // End of literal, stop.
-      break;
-    }
-  }
-  if (HadError)
-    return formToken(tok::unknown, TokStart);
+  // Ask libswift to try and lex a regex literal. If it produces an error (e.g
+  // if it encounters invalid UTF-8, or the literal is unterminated), it will
+  // set ErrStr, and return a suitable EndPtr from which to resume lexing. If it
+  // cannot lex a regex literal, it will return nullptr.
+  const char *ErrStr = nullptr;
+  auto *EndPtr = regexLiteralLexingFn(TokStart, BufferEnd, &ErrStr);
+  if (!EndPtr)
+    return false;
 
+  // Update to point to where we ended regex lexing.
+  assert(EndPtr > TokStart && EndPtr <= BufferEnd);
+  CurPtr = EndPtr;
+
+  if (ErrStr) {
+    diagnose(TokStart, diag::regex_literal_parsing_error, ErrStr);
+    formToken(tok::unknown, TokStart);
```

Comment:
There might be some cases where we emit an error, but still want to form a regex literal token.

String literal has a `FIXME` where even if the end delimiter is missing, we should form a string literal and type check it, so that we can get additional semantic diagnostics, and more importantly, we can get code completion even inside malformed string interpolations. For example, consider `"Name: \(user.#HERE# "`, in this case, end delimiter is missing because it's consumed as the string interpolation, because `)` is missing.  But we still want code completion for `user`.

How about changing `regexLiteralLexingFn` like:

```cpp
const char *Ptr = TokStart;
const char *ErrStr = nullptr;

// Success: status = SUCCESS, advances `Ptr`, ErrStr == nullptr..
// Error, but parsable: status = SUCCESS, advances `Ptr` and populate `ErrStr`. 
// Error, non parsable: status = FAILED, advances `Ptr` and populate `ErrStr`. 
// Total failure: status = FAILED, Ptr == TokStart, ErrStr == nullptr.

bool status = regexLiteralLexingFn(&Ptr, BufferEnd, &ErrStr)
if (ErrStr) {
  diagnose(TokStart, diag::regex_literal_parsing_error, ErrStr);
}
if (Ptr == TokStart) {
  return false;
}
assert(Ptr > TokStart);
formToken(status ? tok::unknown : tok::regex_literal, TokStart);
return false
```

Reply:
@rintaro How does this look?